src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs

```diff
@@ -279,6 +249,29 @@ async Task ProcessInitialEvaluationDataAsync(CancellationToken cancellationToken
                 // Let the exception escape, to unsubscribe data sources.
                 throw;
             }
+
+            (string languageName, string projectFilePath) GetProjectData()
+            {
+                snapshot.Properties.TryGetValue(ConfigurationGeneral.LanguageServiceNameProperty, out string? languageName);
+                snapshot.Properties.TryGetValue(ConfigurationGeneral.MSBuildProjectFullPathProperty, out string? projectFilePath);
+
+                return (string.IsNullOrEmpty(languageName), string.IsNullOrEmpty(projectFilePath)) switch
+                {
+                    (true, false) => throw Fail("Missing language name."),
+                    (false, true) => throw Fail("Missing project file path."),
+                    (true, true) => throw Fail("Missing language name and project file path"),
+                    (false, false) => (languageName, projectFilePath)
+                };
```

Comment:
That's the most interesting way I've seen that done. Works well for 2 entities to check. If it were 3 or more, it would start to get a bit goofy. 😝